File: latest/scholarly_api.py
import requests
from requests.exceptions import RequestException
import json


class InvalidArgumentError(Exception):

  def __init__(self, message):
    super().__init__(message)
    self.name = 'InvalidArgumentError'


class TimeoutError(Exception):

  def __init__(self, message):
    super().__init__(message)
    self.name = 'TimeoutError'


def check_error(response):
  """
    Checks the response for possible errors and raises appropriate exceptions.

    Args:
        response (requests.Response): The response object to be checked for errors.
    """
  if response.status_code == 400:
    raise InvalidArgumentError(
      f'Invalid argument. Status: {response.status_code}')
  if not response.ok:
    raise Exception(f'Failed request. Status: {response.status_code}')


def get_session(profile="default", timeout=None):
  """
    Fetches a new session ID.

    Args:
        profile (str, optional): The profile for Coach.
        timeout (float, optional): The request timeout in seconds.

    Returns:
        dict: The string containing the session ID.
    """
  url = 'https://api.scholarly.repl.co/get-session'
  data = {"profile":profile}
  try:
    response = requests.post(url, json=data, timeout=timeout)
    check_error(response)
    data = response.json()
    return data
  except RequestException as error:
    if error.__class__.__name__ == 'timeout':
      raise TimeoutError(f'getSession Error: {error}')
    raise error


def get_messages(session, timeout=None):
  """
    Fetches the messages for the provided session ID.

    Args:
        session (str): The session ID.
        timeout (float, optional): The request timeout in seconds.

    Returns:
        list: A list containing the messages for the session. Formatted for OpenAI API.
    """
  url = 'https://api.scholarly.repl.co/get-messages'
  data = {"session": session}
  try:
    response = requests.post(url, json=data, timeout=timeout)
    check_error(response)
    return response.json()
  except RequestException as error:
    if error.__class__.__name__ == 'timeout':
      raise TimeoutError(f'getMessages Error: {error}')
    raise error


def coach(prompt, session=get_session(), stream=False, timeout=None):
  """
    Prompt the AI language model Coach.

    Args:
        prompt (str): The prompt to send to Coach .
        session (str): The session ID. Defaults to a new session.
        timeout (float, optional): The request timeout in seconds.
        stream (bool, optional); Whether the response should be streamed. Defaults to False.

    Returns (if not streamed):
        str: The final content returned by coach.    
    
    Yields (if streamed):
        str: The content chunks returned by the coach stream.
    """
  url = 'https://api.scholarly.repl.co/coach'
  data = {"session": session, "prompt": prompt, "stream": stream}

  def stream_response():
    with requests.post(url, json=data, timeout=timeout, stream=True) as response:
        for chunk in response.iter_content(chunk_size=None, decode_unicode=True):
          yield chunk
  
  try:
    if stream:
      return stream_response()
    else:
      response = requests.post(url, json=data, timeout=timeout)
      check_error(response)
      return response.json()
  except RequestException as error:
    if error.__class__.__name__ == 'timeout':
      raise TimeoutError(f'Coach Error: {error}')
    raise error

File: latest/scholarly_api.js
/**
 * Represents an error for invalid arguments passed to a function.
 */
export class InvalidArgumentError extends Error {
    constructor(message) {
      super(message);
      this.name = 'InvalidArgumentError';
    }
  }
  
  /**
   * Represents a custom error class for timeouts.
   */
  export class TimeoutError extends Error {
    constructor(message) {
      super(message);
      this.name = 'TimeoutError';
    }
  }
  
  /**
   * Checks the response for possible errors and throws appropriate exceptions.
   * @param {Response} response - The response object to be checked for errors.
   */
  function checkError(response) {
    if (response.status === 400) {
      throw new InvalidArgumentError(`Invalid argument. Status: ${response.status}`);
    }
    if (!response.ok) {
      throw new Error(`Failed request. Status: ${response.status}`);
    }
  }
  
  /**
   * Fetches the API key with the provided key name.
   * @param {string} keyName - The name of the API key to fetch.
   * @param {number} timeout - The request timeout in milliseconds.
   * @returns {Promise} A promise that resolves to the JSON response containing the API key.
   */
  export async function getKey(keyName, timeout) {
    try {
      const data = '{"keyName":"' + keyName + '"}';
      const response = await fetch('https://api.scholarly.repl.co/get-key', {
        method: 'POST',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: data,
        signal: AbortSignal.timeout(timeout),
      });
  
      checkError(response);
      return await response.json();
    } catch (error) {
      if (error.name === 'AbortError') {
        throw new TimeoutError(`getKey Error: ${error.message}`);
      }
      throw error;
    }
  }
  
  /**
   * Fetches a new session ID.
   * @param {number} timeout - The request timeout in milliseconds.
   * @returns {Promise} A promise that resolves to the JSON response containing the session ID.
   */
  export async function getSession(timeout) {
    const url = 'https://api.scholarly.repl.co/get-session';
    try {
      const response = await fetch(url, { signal: AbortSignal.timeout(timeout)});
      checkError(response);
      const data = await response.json();
      return data;
    } catch (error) {
      if (error.name === 'AbortError') {
        throw new TimeoutError(`getSession Error: ${error.message}`);
      }
      else { throw error; }
    }
  }
  
  /**
   * Fetches the messages for the provided session ID.
   * @param {string} session - The session ID.
   * @param {number} timeout - The request timeout in milliseconds.
   * @returns {Promise} A promise that resolves to the JSON response containing the messages for the session.
   */
  export async function getMessages(session, timeout) {
    const url = 'https://api.scholarly.repl.co/get-messages';
    const data = '{"session":"' + session + '"}';
    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: data,
        signal: AbortSignal.timeout(timeout),
      });
  
      checkError(response);
      return await response.json();
    } catch (error) {
      if (error.name === 'AbortError') {
        throw new TimeoutError(`getMessages Error: ${error.message}`);
      }
      throw error;
    }
  }
  
  /**
   * Fetches the coach stream for the given prompt and session ID, returning an async generator with content chunks.
   * @param {string} prompt - The prompt to send to the coach stream.
   * @param {string} session - The session ID.
   * @param {AbortSignal} signal - The optional AbortSignal for handling cancellations.
   * @returns {AsyncGenerator} An async generator that yields content chunks from the coach stream.
   */
  async function* fetchCoachStream(prompt, session, signal) {
    const response = await fetch("https://api.scholarly.repl.co/coach", {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ prompt: prompt, session: session, stream: true }),
      signal: signal
    });
  
    if (!response.ok) {
      throw new Error(`Error: ${response.status}`);
    }
  
    const reader = response.body.getReader();
    const decoder = new TextDecoder('utf-8');
  
    try {
      while (true) {
        const { value, done } = await reader.read();
        if (done) {
          break;
        }
  
        const content = decoder.decode(value, { stream: true });
        yield content;
  
        if (signal.aborted) {
          break;
        }
      }
    } catch (error) {
      if (error.name === 'AbortError' && reader) {
        await reader.cancel();
      }
      throw error;
    }
  }
  
  /**
   * Fetches a response from Coach.
   * @param {string} prompt - The prompt to send to coach.
   * @param {string} session - The session ID.
   * @param {number} timeout - The request timeout in milliseconds.
   * @returns {Promise} A promise that resolves to the coach response.
   */
  export async function coachResponse(prompt, session, timeout) {
    const url = 'https://api.scholarly.repl.co/coach';
    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ prompt: prompt, session: session, stream: false }),
        signal: AbortSignal.timeout(timeout),
      });
  
      checkError(response);
      return await response.json();
    } catch (error) {
      if (error.name === 'AbortError') {
        throw new TimeoutError(`coachResponse Error: ${error.message}`);
      }
      throw error;
    }
  }
  
  /**
   * Streams content from the coach API given a prompt and sessionID.
   * @param {string} prompt - The prompt to send to coach.
   * @param {string} session - The session ID. Defaults to a new session.
   * @param {function} onContentReceived - Callback function for handling content received.
   * @param {function} onStreamFinished - Callback function for when the stream has finished.
   * @param {function} onError - Callback function for handling errors during streaming.
   * @param {AbortSignal} signal - The optional AbortSignal for handling cancellations.
   */
  export async function coachStream(prompt, session=getSession(null), onContentReceived, onStreamFinished, onError, signal) {
    const abortController = new AbortController();
    const onStopped = (manual) => {
    if (onStreamFinished) {
        onStreamFinished(manual);
    }
    };
      
    try {
        for await (const content of fetchCoachStream(prompt, session, signal ? signal : abortController.signal)) {
          onContentReceived(content);
          if (signal && signal.aborted) {
            abortController.abort();
          }
        }
        onStopped(false);
      }
    catch (err) {
        if (err.name === 'AbortError') {
          onStopped(true);
        } else {
          onError(err);
          onStreamFinished(false);
        }
    }
  };
